igvm/src/c_api.rs

```diff
@@ -0,0 +1,366 @@
+                    if header.write_binary_header(&mut header_binary).is_err() {
+                        result = IgvmApiError::IGVMAPI_INVALID_FILE;
+                    } else {
+                        result = copy_header_data(&header_binary, buf, buf_len);
+                    }
+                } else {
+                    result = IgvmApiError::IGVMAPI_INVALID_PARAMETER;
+                }
+            }
+            IgvmHeaderSection::HEADER_SECTION_DIRECTIVE => {
+                if let Some(header) = (*igvm).directive_headers.get_mut(index as usize) {
+                    let mut header_binary = Vec::<u8>::new();
+                    if header
+                        .write_binary_header(0, &mut header_binary, &mut Vec::<u8>::new())
+                        .is_err()
+                    {
+                        result = IgvmApiError::IGVMAPI_INVALID_FILE;
+                    } else {
+                        result = copy_header_data(&header_binary, buf, buf_len);
+                    }
+                } else {
+                    result = IgvmApiError::IGVMAPI_INVALID_PARAMETER;
+                }
+            }
+            _ => (),
+        }
+    }
+    result
+}
+
+/// Get the associated file data in binary form for the entry with the given
+/// index for a particular section in the parsed IGVM file. This function is
+/// only relevant for headers that refer to a section in the 'file data'
+/// portion of the IGVM file.
+///
+/// If called for an entry that does not contain file data then this function
+/// returns IGVMAPI_NO_DATA.
+///
+/// The caller must allocate a buffer that is large enough to hold the
+/// entire data. The size can be determined by examining the related header
+/// or by calling this function with a buf_len of 0. On return, buf_len
+/// will contain the required size.
+///
+/// On entry, buf_len must be set to the actual size of the buffer pointed
+/// to by buf. On return, buf_len is updated to the actual length.
+///
+/// If the input buffer is too small then this function returns
+/// IGVMAPI_BUFFER_TOO_SMALL and buf_len is set to the required size of
+/// the buffer.
+///
+/// # Safety
+/// This function is exported and designed to be called as a C API. As such
+/// it works with raw pointers and is inherently unsafe due to this.
+#[no_mangle]
+pub unsafe extern "C" fn igvm_get_header_data(
+    igvm: *mut IgvmFile,
+    section: IgvmHeaderSection,
+    index: u32,
+    buf: *mut u8,
+    buf_len: *mut u32,
+) -> IgvmApiError {
+    // Safety: Using pointer from C
+    unsafe {
+        match section {
+            IgvmHeaderSection::HEADER_SECTION_PLATFORM => {}
+            IgvmHeaderSection::HEADER_SECTION_INITIALIZATION => {}
+            IgvmHeaderSection::HEADER_SECTION_DIRECTIVE => {
+                if let Some(header) = (*igvm).directive_headers.get_mut(index as usize) {
+                    let mut file_data = Vec::<u8>::new();
```

Comment:
It's somewhat unfortunate we need this _extra_ buffer. Given that we should be able to make a slice and query it without needing to do this.

Maybe we should add a method to the header types to ask for the size, so this looks cleaner?

Reply:
I've reworked this significantly now. Everything is now managed by handles that hold an instance of both the main file structure and any buffers that have been returned from the query functions. This prevents unnecessary copying and also reduces the amount of unsafe code.